### ISSUE TITLE
Support for multiline string?

### ISSUE THREAD

**Issue by james-em** [glebm/i18n-tasks#383]
Hi,

Sorry in advance if it was already asked. I was wondering if there was a way to prevent this behavior:

```
en:
  admin_mailer:
    profile_approve_email:
      email: >
        <p>Hi</p>
        <p>a new profile is now available to be approved <a href="%{url}">here</a></p>
```

to become

```
en:
  admin_mailer:
    profile_approve_email:
      email: '<p>Hi</p> <p>a new profile is now available to be approved <a href="%{url}">here</a></p>

        '
```

![Screen Shot 2021-06-22 at 12 02 16 PM](https://user-images.githubusercontent.com/63726983/122960717-b0fb4100-d351-11eb-8341-63d4fae08e7b.png)

Thank you !

**Comment by cooljl31** [comment]
Hi! I'm having the same problem running the `i18n-tasks normalize`.

```
-        unsupported_endorsement_type: "Für den Zusatzservice DHL Endorsement wird ein unterstützter 'handling' Wert benötigt. Unterstützte Werte sind: %{supported_values}"
+        unsupported_endorsement_type: 'Für den Zusatzservice DHL Endorsement wird ein unterstützter ' 'handling' ' Wert benötigt. Unterstützte Werte sind: %{supported_values}'
```

I think for that we need an option to disable the conversion `" => '`

**Comment by davidwessman** [comment]
@james-em 

Does it work if you change to use the pipe-operator instead?
```yaml
en:
  admin_mailer:
    profile_approve_email:
      email: |
        <p>Hi</p>
        <p>a new profile is now available to be approved <a href="%{url}">here</a></p>
```

If I do `YAML.load(File.read("./james_example.xml"))` the new lines are already gone, but if I use `|` instead of `>` it works better.

**Comment by otagi** [comment]
I'm using the pipe-operator, and the multi-lines are also concatenated into a single one (without space: `…</p><p>…`) on the target locale when I run e.g. `i18n-tasks translate-missing`.

**Comment by davidwessman** [comment]
> I'm using the pipe-operator, and the multi-lines are also concatenated into a single one (without space: `…</p><p>…`) on the target locale when I run e.g. `i18n-tasks translate-missing`.

Could you add an example?

**Comment by jaredmoody** [comment]
TL;DR this is just a matter of understanding multi-line strings in YAML (which I did not until now), and if you don't want that final newline, use `>-` to strip it out.

Details:

Using just `YAML.load` and `to_yaml`, the following yaml:

```
---
multiline: >
  Hello,
  World
```

Parses to:

`{"multiline"=>"Hello, World\n"}`

Which dumps to:

```
---
multiline: 'Hello, World

  '
```

The line breaks in the middle of the string are ignored but a single final trailing newline is preserved - this is the default block scalar behavior of YAML.  While this may be unexpected when you're not familiar, it's technically correct - this parses back into to the same hash. If you don't want that trailing newline, use YAML's block chomping modifier, `-`:

```
---
multiline: >-
  Hello,
  World
```

This parses to:

`{"multiline"=>"Hello, World"}`

(Note no trailing newline)

Which dumps to:

```
---
multiline: Hello, World
```

Since this gem is doing nothing but using the default YAML behavior, I don't think there's anything that should be changed here and this issue should be closed. If the gem modified the default YAML behavior, that would be unexpected and could break code for users expecting/relying on how YAML works.

Further reading on multiline strings in YAML: https://yaml-multiline.info

**Comment by erikaxel** [comment]
It might be that the underlying issue comes from how the yaml parser works, but according to https://yaml-multiline.info/, the pipe | should keep newlines as is. I would then expect it to do so in a yaml file as well. However we experience that it doesn't, and even worse, i18n-tasks keeps changing between `|` to `>` or to "" in a seemingly random fashion for us. An example:
```
a: | 
  Lorem ipsum dolor sit amet, consectetur
  Lorem ipsum dolor sit amet, consectetur
b: | 
  Lorem ipsum dolor sit amet, consectetur 
  Lorem ipsum dolor sit amet, consectetur 
```
will get changed to
```
a: |
  Lorem ipsum dolor sit amet, consectetur
  Lorem ipsum dolor sit amet, consectetur
b: "Lorem ipsum dolor sit amet, consectetur \nLorem ipsum dolor sit amet, consectetur\n"
```
It can be difficult to see why, but if you check very closely, in `b` there is an extra space at the end of the first line. But I don't see any reason why one should be normalized to "" and the other can keep the `|`. Both forms should be acceptable?
This is one example, but there are more, and nudging out when it keeps one form or another is quite difficult. We wanted to standardize on one type (eg we only do pipes), but that has been impossible. In general, we would prefer to use |, which is very nice when we have long descriptions. It would also be nice if i18n normalize didn't change that.

Maybe this has to be solved upstream, but according to the readme, "normalize" is supposed to: 
```
Sort the keys, and move them to the respective files as defined by [config.write](https://github.com/glebm/i18n-tasks/#multiple-locale-files)
```
Now it does a whole lot more. Perhaps one way to solve it would be to do the sorting and moving without parsing them through yaml?

**Comment by davidwessman** [comment]
Good examples, but I think that is also just taken from the parsing and formatting from YAML itself.
E.g. how would you format a string with two spaces at the end of a line?

When moving around keys i18-tasks parses all files and then writes all of them again.
So we cannot avoid these parsing behaviours if we want to change anything.

**Comment by erikaxel** [comment]
> E.g. how would you format a string with two spaces at the end of a line?

Good question, according to my quick test right now i18n-task would remove that double space if you try to use `|` or `>`. Not sure that is the correct thing to do according to the YAML spec though?

I know it probably is not possible given the technical difficulties, but from a user perspective I think it would have been preferable if i18n-task could either
a) Not convert between `|`, `""` , `''` and `>`.  
b) Allow to set a preferred version for multi-lines.

**Comment by davidwessman** [comment]
Then we would need to override the default behaviour for `Psych`:

```ruby
      example = <<~YAML
        ---
        a: |
          Lorem ipsum dolor sit amet, consectetur
          Lorem ipsum dolor sit amet, consectetur
        b: |
          Lorem ipsum dolor sit amet, consectetur  
          Lorem ipsum dolor sit amet, consectetur
      YAML
      parsed = Psych.safe_load(example)
      expect(Psych.dump(parsed)).to eq(example)
```

```ruby
            expected: "---\na: |\n  Lorem ipsum dolor sit amet, consectetur\n  Lorem ipsum dolor sit amet, consectetur\nb: |\n  Lorem ipsum dolor sit amet, consectetur  \n  Lorem ipsum dolor sit amet, consectetur\n"
            got: "---\na: |\n  Lorem ipsum dolor sit amet, consectetur\n  Lorem ipsum dolor sit amet, consectetur\nb: \"Lorem ipsum dolor sit amet, consectetur  \\nLorem ipsum dolor sit amet, consectetur\\n\"\n"
     
       (compared using ==)
     
       Diff:
       @@ -2,7 +2,5 @@
        a: |
          Lorem ipsum dolor sit amet, consectetur
          Lorem ipsum dolor sit amet, consectetur
       -b: |
       -  Lorem ipsum dolor sit amet, consectetur  
       -  Lorem ipsum dolor sit amet, consectetur
       +b: "Lorem ipsum dolor sit amet, consectetur  \nLorem ipsum dolor sit amet, consectetur\n"
```

**Comment by davidwessman** [comment]
But yes, we do remove trailing spaces, because there was a lot of problems with different versions of libyaml giving different results :/ 

https://github.com/glebm/i18n-tasks/issues/355

**Comment by davidwessman** [comment]
@otagi I never saw your example on this issue.
Seems like we would want the translation api to keep newlines.

But we should move it to a separate issue.

**Comment by davidwessman** [comment]
@erikaxel I added some specs to see how we work with multiline and spaces. I do not find any syntax or setting in Psych or in https://yaml-multiline.info that would help with making sure we always use the same block scalar style.

https://github.com/glebm/i18n-tasks/pull/568

Please add examples to above spec if you find anything that should be handled differently.

**Comment by erikaxel** [comment]
Thank you @davidwessman. Perhaps we could add this caveat to the readme? I think that would be helpful for other people running into this issue thinking it is a fixable bug.

**Comment by davidwessman** [comment]
> Thank you @davidwessman. Perhaps we could add this caveat to the readme? I think that would be helpful for other people running into this issue thinking it is a fixable bug.

Sound reasonable, please feel free to open up a PR for the README 🙂

**Comment by erikaxel** [comment]
Sure, opened one here https://github.com/glebm/i18n-tasks/pull/570.

**Comment by jaredmoody** [comment]
Tangent, but perhaps this gem should not be responsible for formatting/normalizing yaml at all. Seems like formatting could be seen as falling outside the scope of managing translations, and there are other libraries like [yamlfmt](https://github.com/google/yamlfmt) out there. It looks like yamlfmt handles preserving multiline formatting as well.

**Comment by davidwessman** [comment]
The problem is that the gem needs to read and write all files when normalizing, so while we could use something else for formatting - we will actually rewrite everything.
